From 65b9e84b89e5826491da032a748d7b5c6735e945 Mon Sep 17 00:00:00 2001
From: Sebastien PASDELOUP <sebastien.pasdeloup-ext@st.com>
Date: Mon, 5 Jun 2023 09:32:22 +0200
Subject: [PATCH 188/841] feat(stm32mp1): add stm32_get_uid_otp

Add a generic function to get the unique device id (UID)
from the OTP.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git bca57856edfcb94cf2ffab1e5fce0e1691e3f5b0 feat(stm32mp1): add stm32_get_uid_otp]
Signed-off-by: Patrick Delaunay <patrick.delaunay@foss.st.com>
Change-Id: Id15dbb6668b4bafe44e86c33e76ec64a3c982387
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/255795
Reviewed-by: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Reviewed-by: Nicolas LE BAYON <nicolas.le.bayon@st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/common/include/stm32mp_common.h |  2 ++
 plat/st/common/stm32mp_common.c         | 26 +++++++++++++++++++++++++
 plat/st/stm32mp1/stm32mp1_usb_dfu.c     | 18 +----------------
 3 files changed, 29 insertions(+), 17 deletions(-)

diff --git a/plat/st/common/include/stm32mp_common.h b/plat/st/common/include/stm32mp_common.h
index a90a48fea..d470f1075 100644
--- a/plat/st/common/include/stm32mp_common.h
+++ b/plat/st/common/include/stm32mp_common.h
@@ -56,6 +56,8 @@ int stm32_get_otp_index(const char *otp_name, uint32_t *otp_idx,
 int stm32_get_otp_value(const char *otp_name, uint32_t *otp_val);
 int stm32_get_otp_value_from_idx(const uint32_t otp_idx, uint32_t *otp_val);
 int stm32_lock_enc_key_otp(void);
+/* update UID_WORD_NB array */
+int stm32_get_uid_otp(uint32_t uid[]);
 
 /* Get IWDG platform instance ID from peripheral IO memory base address */
 uint32_t stm32_iwdg_get_instance(uintptr_t base);
diff --git a/plat/st/common/stm32mp_common.c b/plat/st/common/stm32mp_common.c
index 3c3813661..6c32bb0e4 100644
--- a/plat/st/common/stm32mp_common.c
+++ b/plat/st/common/stm32mp_common.c
@@ -237,6 +237,32 @@ int stm32_lock_enc_key_otp(void)
 	return 0;
 }
 
+int stm32_get_uid_otp(uint32_t uid[])
+{
+	uint8_t i;
+	uint32_t otp;
+	uint32_t len;
+
+	if (stm32_get_otp_index(UID_OTP, &otp, &len) != 0) {
+		ERROR("BSEC: Get UID_OTP number Error\n");
+		return -1;
+	}
+
+	if ((len / __WORD_BIT) != UID_WORD_NB) {
+		ERROR("BSEC: Get UID_OTP length Error\n");
+		return -1;
+	}
+
+	for (i = 0U; i < UID_WORD_NB; i++) {
+		if (bsec_shadow_read_otp(&uid[i], i + otp) != BSEC_OK) {
+			ERROR("BSEC: UID%u Error\n", i);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
 #if  defined(IMAGE_BL2)
 static void reset_uart(uint32_t reset)
 {
diff --git a/plat/st/stm32mp1/stm32mp1_usb_dfu.c b/plat/st/stm32mp1/stm32mp1_usb_dfu.c
index 0fe2d24ae..305ebf13e 100644
--- a/plat/st/stm32mp1/stm32mp1_usb_dfu.c
+++ b/plat/st/stm32mp1/stm32mp1_usb_dfu.c
@@ -176,31 +176,15 @@ static void stm32mp1_get_string(const char *desc, uint8_t *unicode, uint16_t *le
  */
 static void update_serial_num_string(void)
 {
-	uint8_t i;
 	char serial_string[SIZ_STRING_SERIAL + 2U];
 	/* serial number is set to 0 */
 	uint32_t deviceserial[UID_WORD_NB] = {0U, 0U, 0U};
-	uint32_t otp;
-	uint32_t len;
 	uint16_t length;
 
-	if (stm32_get_otp_index(UID_OTP, &otp, &len) != 0) {
-		ERROR("BSEC: Get UID_OTP number Error\n");
+	if (stm32_get_uid_otp(deviceserial) != 0) {
 		return;
 	}
 
-	if ((len / __WORD_BIT) != UID_WORD_NB) {
-		ERROR("BSEC: Get UID_OTP length Error\n");
-		return;
-	}
-
-	for (i = 0; i < UID_WORD_NB; i++) {
-		if (bsec_shadow_read_otp(&deviceserial[i], i + otp) !=
-		    BSEC_OK) {
-			ERROR("BSEC: UID%d Error\n", i);
-			return;
-		}
-	}
 	/* build serial number with OTP value as in ROM code */
 	snprintf(serial_string, sizeof(serial_string), "%08X%08X%08X",
 		 deviceserial[0], deviceserial[1], deviceserial[2]);
-- 
2.47.2

